packages/core/src/internal/proc.js

```diff
@@ -350,31 +351,33 @@ export default function proc(
-
-      // declarative effects
-      : (data = asEffect.take(effect))          ? runTakeEffect(data, currCb)
-      : (data = asEffect.put(effect))           ? runPutEffect(data, currCb)
-      : (data = asEffect.all(effect))           ? runAllEffect(data, effectId, currCb)
-      : (data = asEffect.race(effect))          ? runRaceEffect(data, effectId, currCb)
-      : (data = asEffect.call(effect))          ? runCallEffect(data, effectId, currCb)
-      : (data = asEffect.cps(effect))           ? runCPSEffect(data, currCb)
-      : (data = asEffect.fork(effect))          ? runForkEffect(data, effectId, currCb)
-      : (data = asEffect.join(effect))          ? runJoinEffect(data, currCb)
-      : (data = asEffect.cancel(effect))        ? runCancelEffect(data, currCb)
-      : (data = asEffect.select(effect))        ? runSelectEffect(data, currCb)
-      : (data = asEffect.actionChannel(effect)) ? runChannelEffect(data, currCb)
-      : (data = asEffect.flush(effect))         ? runFlushEffect(data, currCb)
-      : (data = asEffect.cancelled(effect))     ? runCancelledEffect(data, currCb)
-      : (data = asEffect.getContext(effect))    ? runGetContextEffect(data, currCb)
-      : (data = asEffect.setContext(effect))    ? runSetContextEffect(data, currCb)
-      : /* anything else returned as is */        currCb(effect)
-    )
+    if (is.promise(effect)) {
+      return resolvePromise(effect, currCb)
+    } else if (is.iterator(effect)) {
+      return resolveIterator(effect, effectId, meta, currCb)
+    } else if (effect && effect[IO]) {
+      const { type, payload } = effect
+      // prettier-ignore
+      return (type === EFFECT_TYPES.TAKE)      ? runTakeEffect(payload, currCb)
+      : (type === EFFECT_TYPES.PUT)            ? runPutEffect(payload, currCb)
+      : (type === EFFECT_TYPES.ALL)            ? runAllEffect(payload, effectId, currCb)
+      : (type === EFFECT_TYPES.RACE)           ? runRaceEffect(payload, effectId, currCb)
+      : (type === EFFECT_TYPES.CALL)           ? runCallEffect(payload, effectId, currCb)
+      : (type === EFFECT_TYPES.CPS)            ? runCPSEffect(payload, currCb)
+      : (type === EFFECT_TYPES.FORK)           ? runForkEffect(payload, effectId, currCb)
+      : (type === EFFECT_TYPES.JOIN)           ? runJoinEffect(payload, currCb)
+      : (type === EFFECT_TYPES.CANCEL)         ? runCancelEffect(payload, currCb)
+      : (type === EFFECT_TYPES.SELECT)         ? runSelectEffect(payload, currCb)
+      : (type === EFFECT_TYPES.ACTION_CHANNEL) ? runChannelEffect(payload, currCb)
+      : (type === EFFECT_TYPES.FLUSH)          ? runFlushEffect(payload, currCb)
+      : (type === EFFECT_TYPES.CANCELLED)      ? runCancelledEffect(payload, currCb)
+      : (type === EFFECT_TYPES.GET_CONTEXT)    ? runGetContextEffect(payload, currCb)
+      : (type === EFFECT_TYPES.SET_CONTEXT)    ? runSetContextEffect(payload, currCb)
+      : kThrow(`runEffect(effect...) invalid effect type: ${type}`)
```

Comment:
`kThrow('runEffect(effect...) invalid effect type: ${type}')` could be thought as a fallback `runCustomEffect`. 
Once #1479 is merged, I think we may use `env.runCostumEffect` here, that will allow running user-defined effects.